d3b-center/docker-terraform
# Create an image for running deployments with Terraform on AWS

## Overview

Manipulating infrastructure is hard when developers with various experience levels and machines collaborate on a single project. 

For example, imagine a scenario where:

* Rocky is using Terraform A.A.A installed via Homebrew on his Macbook.
* Alex is using Terraform B.B.B installed via the HashiCorp Linux repository on Ubuntu.
* Someone created the project's Terraform state with Terraform C.C.C.

What happens when Rocky and Alex are both manipulating the Terraform state? What version of Terraform is installed CI?

The solution is to ensure that everyone uses the same version of Terraform.

However, think about developers new to the app. First, they have to install Terraform and the AWS CLI on their system. Then, they have to make sure their version of Terraform aligns with the project's, or they run the risk of corrupting the state file. Finally, they have to pin these packages so they don't accidentally get updated. And, what happens if they're working on two Terraform projects, each with different state versions?

This pull request aims to solve these issues by creating a container image that contains a standard installation of Terraform and the AWS CLI. Now, the new developer can check out the project's codebase onto their development machine, requiring only Docker and Docker Compose installed as prerequisites.

Resolves https://d3b.atlassian.net/browse/DEVOPS-513

## Notes

I've designed this image [in the past](https://github.com/azavea/docker-terraform) by:

* Starting with the [`hashicorp/terraform`](https://hub.docker.com/r/hashicorp/terraform/) base image.
* Adding supporting packages like the AWS CLI via `apk` or `pip`.

I would've taken that approach here, but I wanted to use the AWS CLI v2 to take advantage of the new [AWS SSO integration](https://docs.aws.amazon.com/singlesignon/latest/userguide/integrating-aws-cli.html). Unfortunately, `hashicorp/terraform` is based on Alpine Linux, and Amazon doesn't compile the AWS CLI v2 against musl libc. A [potential solution](https://github.com/aws/aws-cli/issues/4685#issuecomment-615872019) involves patching glibc into Alpine Linux, but I'd consider that more of a hack.

This snafu exposes a much larger problem with OS packaging in general. Take a real-life example I ran into a couple of months ago: We're developing a Python app and basing our image off `python:3.x.x:slim`. Our Python app needs various shared libraries, like GDAL. The problem is, we're using PostgreSQL 12, support for which wasn't [patched into](https://github.com/OSGeo/gdal/blob/v2.4.3/gdal/NEWS#L104-L106) GDAL until v2.4.3. Debian buster [only builds](https://packages.debian.org/buster/gdal-bin) v2.4.0. So, you end up downloading the source files for GDAL v2.4.3, compiling it yourself as part of the container build, and doing your best to produce the same binary that [Salsa would](https://salsa.debian.org/debian-gis-team/gdal/-/blob/master/debian/rules#L120-171).  

The result is worse than "well, it works on my machine." Now, "well, it works on my machine _and_ my compilation of GDAL."

Alexander Bantyev from Serokell [put it eleoquently](https://serokell.io/blog/what-is-nix):

> How often do you find yourself in a situation where something builds and works on your machine, but doesn’t build on CI or fails catastrophically in production?

> In many cases, those problems are a symptom of something most developers face: lack of reproducibility. Your code depends on an environment variable at compile time, or significantly changes behavior with different versions of a dependency.

> Such problems are usually hard to diagnose and even harder to fix. However, there is a tool that can help you solve those issues – Nix.

At a high level, all we need to do is tell Nix, "I want this version of Terraform, and this version of the AWS CLI," and Nix will make everything work. Nix flakes are similar to a `package.json` file. You specify all your dependencies in a `flake.nix` file, and then Nix generates `flake.lock`, akin to `package.json.lock`.

The good:

* Nix can produce _lean_ container images. Our Nix image is smaller than the Alpine-based `hashicorp/terraform` image + [`amazon/aws-cli`](https://hub.docker.com/r/amazon/aws-cli) combined.
* Our expression in `flake.nix` is simple, readable, and extendable. We're not pulling in anything via `curl`.
* Our build is always reproducible. Hashicorp can't update our base image and remove any underlying functionality.

The bad:

* Nix is new. The language (and functional programming paradigm) is a considerable obstacle for new developers.
* We're relying on the public binary cache. If the cache stops serving packages built against our pinned version of [NixOS/nixpkgs](https://github.com/nixos/nixpkgs), we'll have to compile everything from source, which increases build times.
* I haven't figured out a way to pass arguments to Nix flakes, so we have to declare and maintain separate Nix flakes for each version of Terraform we wish to support, rather than relying more heavily on the GitHub Actions build matrix. We can improve on this in the future.

## Testing Instructions

Follow the [instructions](https://github.com/d3b-center/docker-terraform/tree/jrb/bootstrap#testing) in the README to build the image:

```console
$ CI=1 TERRAFORM_VERSION=1.1.6 ./scripts/cibuild
```

You can run tests with `scripts/test`:

```console
$ CI=1 TERRAFORM_VERSION=1.1.6 ./scripts/test
+ set -u
+ [[ ./scripts/test == \.\/\s\c\r\i\p\t\s\/\t\e\s\t ]]
+ [[ '' == \-\-\h\e\l\p ]]
+ docker run --rm ghcr.io/d3b-center/terraform:1.1.6 -version
Terraform v1.1.6
on linux_amd64
+ docker run --rm --entrypoint aws ghcr.io/d3b-center/terraform:1.1.6 --version
aws-cli/2.4.9 Python/3.9.10 Linux/5.10.76-linuxkit source/x86_64 prompt/off
```

Finally, see the GitHub Actions workflow: https://github.com/d3b-center/docker-terraform/runs/5293577908.
